src/network/dn_comp.c

```diff
@@ -113,29 +115,25 @@ int dn_comp(const char *src : itype(_Nt_array_ptr<const char>),
 	int i, j, n, m=0, offset, bestlen=0, bestoff;
 	unsigned char lens _Checked[127];
 	size_t l;
-	// TODO: bounds-safe interface for the 1st parameter of strnlen appears incorrect. It should
-	// accept an Nt_array_ptr with count(0).
-	_Unchecked {
-		l = strnlen((const char *)src, 255);
-	}
-	_Nt_array_ptr<const char> src_l : count(l) = 0;
-	_Unchecked {
-		src_l = _Assume_bounds_cast<_Nt_array_ptr<const char>>(src, count(l));
-	}
-	if (l && src_l[l-1] == '.') l--;
+	_Unchecked { l = strnlen((const char *)arg_src, 255); }
```

Comment:
You can implement `strnlen` safely for an `nt_array_ptr` argument in Checked C without using `memchr`.   The requirements of `memchr` should not affect the bounds-safe interface for `strnlen`.    The use of `memchr` is a performance optimization.   This means that we should be able to allow bounds widening for `strnlen`, just like we allow bounds widening for `strlen`.

The use of `memchr(s, 0, n)` is semantically safe when called with an `nt_array_ptr` because it will stop when it hits the null terminator.  However Checked C does not have an expression enough type system to statically express the safety. 
 The allowed type of the first argument would depend on whether the 2nd argument is 0 or not.  Because Checked C cannot express the safety of the call to`memchr`, you'd have to wrap the use in `strnlen` in an unchecked block.

strnlen is an interesting function because it actually could have two different bounds-safe interfaces, depending on whether the function argument is `nt_array_ptr` or an `array_ptr`.
- It can be called on an `nt_array_ptr<char>` with bounds of count(0).
-  It can be called on an `array_ptr<char>` with bounds of n.  Interestingly, in this case, Checked C could prove that the use of `memchr` is safe.

We have run into other cases where a function could be used on either `array_ptr` or an `nt_array_ptr` with differing bounds.  Our recommendation is to just create two different copies of the function with the different bounds.